src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -517,89 +477,117 @@ internal unsafe ValueTask ConnectAsync(CancellationToken cancellationToken = def
+                if (_remoteEndPoint is IPEndPoint ipEndPoint)
+                {
+                    Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
+                    MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, ipEndPoint);
+                    targetHost = _state.TargetHost ?? ((IPEndPoint)_remoteEndPoint).Address.ToString();
+                    port = ((IPEndPoint)_remoteEndPoint).Port;
 
-                // We don't have way how to set separate SNI and name for connection at this moment.
-                // If the name is actually IP address we can use it to make at least some cases work for people
-                // who want to bypass DNS but connect to specific virtual host.
-                if (!dnsHost.Equals(_state.TargetHost, StringComparison.InvariantCultureIgnoreCase) && !string.IsNullOrEmpty(_state.TargetHost))
+                }
+                else if (_remoteEndPoint is DnsEndPoint dnsEndPoint)
                 {
-                    targetHost = _state.TargetHost!;
-                    if (IPAddress.TryParse(dnsHost, out IPAddress? address))
+                    port = dnsEndPoint.Port;
+                    string dnsHost = dnsEndPoint.Host!;
+
+                    // We don't have way how to set separate SNI and name for connection at this moment.
+                    // If the name is actually IP address we can use it to make at least some cases work for people
+                    // who want to bypass DNS but connect to specific virtual host.
+                    if (!dnsHost.Equals(_state.TargetHost, StringComparison.InvariantCultureIgnoreCase) && !string.IsNullOrEmpty(_state.TargetHost))
                     {
-                        // This is form of IPAddress and _state.TargetHost is set to different string
-                        Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
-                        MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, new IPEndPoint(address, port));
+                        targetHost = _state.TargetHost!;
+                        if (IPAddress.TryParse(dnsHost, out IPAddress? address))
+                        {
+                            // This is form of IPAddress and _state.TargetHost is set to different string
+                            Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
+                            MsQuicParameterHelpers.SetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS, new IPEndPoint(address, port));
+                        }
+                        else
+                        {
+                            IPAddress[] addresses = Dns.GetHostAddressesAsync(dnsHost, cancellationToken).GetAwaiter().GetResult();
+                            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Why is this needed?

Reply:
Based on @wfurt comment: https://github.com/dotnet/runtime/pull/71428#discussion_r909682568, on some platforms (e.g. Linux) the token is not processed by the resolution. I'm only shuffling code here.